android/app/src/main/kotlin/coop/polypoly/polypod/updatenotification/UpdateNotification.kt

```diff
@@ -5,23 +5,63 @@ import coop.polypoly.polypod.Preferences
 import coop.polypoly.polypod.R
 
 class UpdateNotification(private val context: Context) {
-    val id = context.resources.getInteger(R.integer.update_notification_id)
+    companion object {
+        data class MockData(var id: Int? = null)
+        val mockData = MockData()
+    }
+
+    private enum class State {
+        NOT_SEEN,
+        PUSH_SEEN,
+        ALL_SEEN;
+
+        companion object {
+            fun parse(s: String?) =
+                s.let { name -> values().firstOrNull { it.name == name } }
+        }
+    }
+
+    val id = mockData.id
+        ?: context.resources.getInteger(R.integer.update_notification_id)
+
     val title = context.getString(R.string.update_notification_title)
+
     val text = context.getString(R.string.update_notification_text)
+
     val pushDelay =
         context.resources.getInteger(R.integer.update_notification_push_delay)
 
-    val inAppNotificationSeen: Boolean
-        get() = id == 0 || Preferences.getSeenInAppNotificationId(context) == id
+    val showPush: Boolean
+        get() = state == State.NOT_SEEN
+
+    val showInApp: Boolean
+        get() = state != State.ALL_SEEN
+
+    private var state: State = loadLastState()
+        set(value) {
+            field = value
+            Preferences.setLastNotification(context, id, state.name)
+        }
+
+    private fun loadLastState(): State {
+        val (lastId, lastState) = Preferences.getLastNotification(context)
+        return when (id) {
+            0 -> State.ALL_SEEN
+            lastId -> State.parse(lastState) ?: State.ALL_SEEN
+            else -> State.NOT_SEEN
+        }
+    }
+
+    fun onStartup() = onPushSeen()
 
-    val pushNotificationSeen: Boolean
-        get() = id == 0 || Preferences.getSeenPushNotificationId(context) == id
+    fun onFirstRun() = onInAppSeen()
```

Comment:
But if they are the same, why the different names?

Reply:
They are different _events_, it's just that (currently), the same thing happens for both events.

Reply:
Well, different *phasers*, since they occur at different *phases* in the lifecycle of a notification... You know I'm against code duplication, but this might be one in which it would actually make sense. It's not that onFirstRun needs to invoke the onInAppSeen functionality, it's just that they (currently) do the same.

Reply:
Yeah, I had duplication there originally, guess I'll bring it back! If the code confused you, it's gonna confuse more people later I suppose.